Adds SAML SP

JPrevost opened the pull request:
Why are these changes being introduced:

* A SAML SP is method we use to interact with MIT Touchstone authentication

Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/GDT-121

How does this address that need:

* Borrowing heavily from our previous application `bd_auth`, this implements a SAML SP via `python3-saml`
* a new make command `make app-bash` was created to streamline local development of this application while avoiding a problematic dependency
* To test these changes locally
  * should create an SP cert/key pair and extract them to single lines with no spaces and set in your local `.env` file (see README for the env values to set these to)
  * start the local IdP (see README)
  * start a local shell for the app with `make app-bash`
  * start the local dev server in the shell from the previous state with `make run-dev`
  * access file that should allow for redirects such as `http://localhost:5000/?cdn_resource=https://cdn.libraries.mit.edu/fakefile`. You should be prompted to authenticate (credentials in README) and then redirect to a "file not found" page for the CDN. You cannot access actually restricted files in the CDN when running locally like this as we pass JWT tokens as domain cookies which cannot be read. You _could_ set a valid domain cookie by modifying /etc/hosts and running your local dev environment on something like `totallyrealserver.libraries.mit.edu`. You'd need to pull the appropriate JWT secret from SSM.
  * See the tests for other URL patterns you may want to try (like leaving out the ?cdn_resource query param or setting it to a value that we won't redirect to for security reasons, etc)

What is missing:

* unit tests for SAML interactions. We should add unit tests for the auth.py module, but for now we need to manually test in our staging environment before moving code to production. As unit tests won't test against live Touchstone, this manual test will remain necessary even after we have automated tests in place.

Document any side effects to this change:

* Development is now expected to be in docker as libxmlsec1 dependency is complicated (at best) to get installed reliably locally (the upstream libraries are slowly working through the bugs and hope to have fixes in April 2024 which should allow us to move back to non-dockerized development if preferred)
* mypy linter was removed as we aren't using types in this application

JPrevost committed:
Adds SAML SP

Why are these changes being introduced:

* A SAML SP is method we use to interact with MIT Touchstone
  authentication

Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/GDT-121

How does this address that need:

* Borrowing heavily from our previous application `bd_auth`, this
  implements a SAML SP via `python3-saml`
* a new make command `make app-bash` was created to streamline local
  development of this application while avoiding a problematic
  dependency

What is missing:

* unit tests. We need to set this app aside for a bit, but want to push
  forward with the Touchstone registration so we will eliminate that
  delay once we return to this. We should add unit tests to this work
  when we move forward with implementing the business logic.

Document any side effects to this change:

* Development is now expected to be in docker as libxmlsec1 dependency
  is complicated (at best) to get installed reliably locally
* mypy linter was removed as we aren't using types in this application

JPrevost committed:
Install libxmlsec1 in CI

JPrevost committed:
Install libxmlsec1 on heroku

See https://elements.heroku.com/buildpacks/heroku/heroku-buildpack-apt
for more information on how this works

JPrevost committed:
Local IdP

JPrevost committed:
Core business logic for setting JWT token

Why are these changes being introduced:

* Users directed to this application for authentication will be
  redirected to Touchstone and then be granted a JWT token and
  redirected back to the originally requested resource

Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/GDT-122
* https://mitlibraries.atlassian.net/browse/GDT-101

How does this address that need:

* Accessing cdn auth with no cdn_resource parameter or a cdn_resource
  parameter that is not in one of our CDNs is rejected as invalid
* A user with an invalid, expired, or missing JWT token is redirected
  to Touchstone
* Valid users coming back from Touchstone will have a valid JWT token
  set as a domain cookie
* Users with a valid JWT token and valid cdn_resource are redirected to
  the resource (which is then intercepted by the cdn auth lambda to
  validate the token before sending them the requested file)
* A preconfigued local IdP can be run in a container for development
  purposes (only SP key/cert need to be generated manually and set in
  .env prior to running `make app-bash`)

Document any side effects to this change:

* Adding tests for the SAML reponse handling are still required, but
  tests exist for all of the core business logic. This has been opened
  as https://mitlibraries.atlassian.net/browse/GDT-250. For now, it is
  expected that we will manually confirm SAML functionality in our
  staging environment prior to moving changes into production.

ghukill commented on README.md:
```diff
@@ -40,17 +40,115 @@ Not all links are publicly accessible.
 - (optional) run prod-mode server locally: `make run-prod`
   - access localhost:8000 and localhost:8000/ping
+
+## Required ENV
```
Minor nit that this markdown renders a little funky ([github rendered form](https://github.com/MITLibraries/cdn-auth-geo/tree/GDT-121-saml-sp?tab=readme-ov-file#required-env)).  Took me awhile to build all the correct env vars as the newlines appear to be dropped.  But seeing it here in the code view, I see the intent and how it would have been easier to discern.

That said, had all the info I needed to get it up and running.

ghukill replied:
With new render, easy to parse!

ghukill commented on README.md:
```diff
@@ -40,17 +40,115 @@ Not all links are publicly accessible.
-- pipenv in your python 3.12 install: `pip install pipenv`
-  - note: failing to do this may not result in a noticeable problem, but this is still good practice as it will ensure you are using a `pipenv` within the correct version of python instead of one from your default version. Do this every time you install a new python version.
-- note: a python virtual environment will be created via `pipenv` when following the next steps if one does not yet exist. If you run into issues at any point, you can remove that virtualenv with `pipenv --rm` from the project root and start over.
+This application expects to be developed in docker. [There are dependency issues with libxmlxec1 on macos](https://github.com/SAML-Toolkits/python3-saml/issues/356) at this time and rather than each of us fighting with that, we will develop inside the docker container so it can be solved programatically and repeatably.
```
_Before_ I read this and the `make app-bash` approach, I struggled to get things installed as well.  Ultimately I did get it installed, where [this github issue workaround](https://github.com/xmlsec/python-xmlsec/issues/254#issuecomment-1612005910) worked well for me.

Related, I also needed to `pip install setuptools`, as it sounds like `disutils` does not ship with python 3.12+.  

With those two things done, `make install` worked as expected, and I'm able to run `make run-dev` and fire up the flask app locally (outside of a docker container).

All that said, I think the docker approach also works well!  Nice to have options for developer's choice.

JPrevost replied:
Glad (sort of) you were able to get it running locally in addition to trying out the docker approach. The upstream requirement issues are starting to resolve themselves and this should be much more easily installable locally soon(TM).

ghukill commented on README.md:
```diff
@@ -40,17 +40,115 @@ Not all links are publicly accessible.
+Note: See [our dev docs](https://mitlibraries.github.io/guides/authentication/touchstone_saml.html#generating-a-self-signed-certificate-for-touchstone) for information on how to generate SP key/cert. They should be unique for each deployment and backed up to our shared LastPass.
+
+`SP_ACS_URL` = route in this app that handles the response from IDP. domain name of app + /saml/?acs
+`SP_CERT` = obtained from self signed cert generated for this app. Note: remove all spaces/linebreaks as well as the "BEGIN" and "END" lines from file for ENV setting.
+`SP_ENTITY_ID` = domain name of app + /saml
+`SP_KEY` = obtained from self signed key generated for this app
```
Just noting that `SP_CERT` and `SP_KEY` were **multiline** after I ran through the -- very helpful -- instructions in page linked above.  I realized I wasn't immediately sure how to set a multiline env var in a `.env` file.  Probably better ways to do it, but I popped them into python as a multiline string, and then could access the string variable which included the `\n` characters, which allowed me to set them like:

```
SP_CERT="-----BEGIN CERTIFICATE-----\nabc123...\ndef456...\n-----END CERTIFICATE-----"
SP_KEY="-----BEGIN PRIVATE KEY-----\nabc123...\ndef456...\n-----END PRIVATE KEY-----"
```

JPrevost replied:
Yeah this should likely be clarified better in the docs... I just manually remove the line breaks to create a single line string. I feel like a `make stringify_cert_or_key` command that can spit out what works from an input key or cert would be useful. I've just been too lazy to write one.

ghukill commented on cdnauth/debug.py:
```diff
@@ -0,0 +1,12 @@
+bp = Blueprint("debug", __name__, url_prefix="/debug")
+
+
+@bp.route("/")
+@login_required
+def item():
+    return render_template("debug.html")
```
Dig it!  This is really handy to have.

JPrevost replied:
There is an interesting side effect the way this works. The cookie setting only happens in the `cdn` module instead of the `auth` module so while we can debug the session like this, we can't actually debug the cookies because they don't get set. I think eventually we should see if we can move the cookie setting logic into `auth`. That added enough complexity to not include that refactor in this initial version because you set cookies as part of the response which is not the domain if `auth` so I left it alone for now.

ghukill commented on cdnauth/auth.py:
```diff
@@ -0,0 +1,119 @@
+    return json_settings
+
+
+def prepare_flask_request(request):
```
Feeling like maybe a docstring here would be helpful.  I had to poke around a bit to understand that the variable `req` which is set from the output of this function (around line 76), is a dictionary in the expected form that the `python3-saml` class `OneLogin_Saml2_Auth` expects for `init`.  

Otherwise, I wondered a bit why we'd collect things from the `request` object that are readily available.

ghukill commented on cdnauth/auth.py:
```diff
@@ -0,0 +1,119 @@
+    saml_settings = load_saml_settings()
+    req = prepare_flask_request(request)
+    auth = OneLogin_Saml2_Auth(req, saml_settings)
+    errors = []
```
Don't think this `errors = []` is needed?  Believe that `errors = auth.get_errors()` below will init the list, and it's only used under that logic branch.

ghukill commented on cdnauth/auth.py:
```diff
@@ -0,0 +1,119 @@
+        auth.process_response()
+        errors = auth.get_errors()
+        if not auth.is_authenticated():
+            # TODO: return something helpful to the user.
+            # That said, this should never happen.
+            pass
```
Is this because the lambda edge before this would ensure that any requests that make it this far are authenticated?  Or am I thinking about this wrong?

I'd propose one of two things given this `TODO`:
1. a docstring for the route as a whole, explaining this situation where a non-authenticated request makes it this far
2. maybe raise an explicit exception here making it clear to those just reading code that this state really shouldn't be encountered

JPrevost replied:
I non-authenticated request should _never_ get this far because Touchstone does not send non-authenticated users back here... if they fail authentication they stay in Touchstone... and if they are _not_ authenticated we send them _to_ Touchstone. This code block is from the example python3-saml provides so I suspect it was just to show what is going on more than making sense to actually do anything. I'll try to make it a lot clearer in the code what is going on. I suspect maybe just logging an exception if this block ever gets run so we'll know I was completely wrong and we need to write code or not :)

ghukill commented on cdnauth/auth.py:
```diff
@@ -0,0 +1,119 @@
+            session["samlSessionIndex"] = auth.get_session_index()
+            return redirect(request.form["RelayState"])
+        else:
+            print("Errors: %s", errors)
+            print("Last error reason: %s", auth.get_last_error_reason())
```
Unsure if it's something we want to tackle now, but we might want to consider moving these to logging statements.  Perhaps the overhead here is setting up logging globally for the full Flask app.

JPrevost replied:
Oh god I have no idea why these are print statements. I suspect I was debugging and forgot to clean up after I got it all working. Oops :)

ghukill commented on cdnauth/cdn.py:
```diff
@@ -0,0 +1,63 @@
+        return "No download URL present"
+
+
+# valid_redirect(url) ensures the redirect is to a domain we trust to prevent
+# our app being used to trick users into clicking on malicious links
+def valid_redirect(url):
+    domain = urlparse(url).netloc
+    if domain in valid_domains():
+        return True
+    return False
```
Small nit: these could be `"""...."""` docstrings for the function.  Same below for `valid_domains()`.

JPrevost replied:
That's good input. I just read the PEP on that and agree it makes sense to move this into the method as a docstring . It still feels backwards to document _inside_ the method instead of before the method as Python is not how I think :)

ghukill commented on cdnauth/cdn.py:
```diff
@@ -0,0 +1,63 @@
+    return False
+
+
+# valid_domains() is a list of domains we trust for redirect purposes
+def valid_domains():
+    return [
+        "cdn.libraries.mit.edu",
+        "cdn.dev1.mitlibrary.net",
+        "cdn.stage.mitlibrary.net",
+    ]
```
Might propose setting this as a constant at the top of the file, something along the lines of:

```python
VALID_DOMAINS = [...]
```

Or, even as a property on the `Config` class (probably only if accessed from multiple files/modules).

JPrevost replied:
I like moving it to `Config` so we could set it but allow overriding via ENV. It's actually convenient in development to allow redirects to `example.com` (or similar) to allow focusing on business logic and not _actually_ invoking a CDN call.

ghukill commented on cdnauth/__init__.py:
```diff
@@ -3,26 +3,27 @@
+elif flask_env == "testing":
     app.config.from_object("cdnauth.config.TestingConfig")
+else:
+    app.config.from_object("cdnauth.config.Config")
```
This is interesting!  Never seen `app.config.from_object(...)`.  But I dig it.  Seems like it streamlines updating an object with worrying about getter/setter considerations.

JPrevost replied:
I must admit I copied that from our old `bd_auth` app (which I also wrote but don't remember doing) and I suspect strongly that it copied it from `ebooks` https://github.com/MITLibraries/ebooks/blob/main/ebooks/__init__.py#L18-L23. It's convenient that the various configs inherit from the base one so you can read everything from env by default but in testing just hard code the values you want to be consistent in a fairly clean way. You only have to set the differences from the base is another way to say that I guess.

ghukill commented on cdnauth/templates/download.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta http-equiv="refresh" content="5; url={{ cdn_resource }}" />
```
Probably a silly question, but does this `<meta>` tag force a refresh of the page, with that CDN link as the URL, thereby prompting a download?  If so, cool!

Also, for people like me unfamiliar with this behavior, might be nice to have a small HTML comment above this that notes this line should force a page refresh.

JPrevost replied:
That's good input. I think everyone in EngX would immediately understand what this does because we live on the HTML side so much more than DataEng. I'll see if I can find a way to better document that (possibly with a comment to link to external docs that explain it clearly). For now it has a 5 second delay before the refresh... once we launch we'll likely set this to an immediate redirect

ghukill replied:
Thanks for adding this.  And, while unsure if necessary in code, thanks for the context about this template page serving as a way to navigate away from the Touchstone page; makes sense.

ghukill reviewed (approved):
Overall, I think it looks good!  Easy to follow and poke around.  Also like the blueprints approach.

I left a few comments: a couple of format-y nit picks, some notes on getting it up and running locally for me, and a couple suggestsions of docstrings.  But all optional, nothing blocking.

I see in the PR comments that typing is not applied to this repository.  If we _did_ want to go back and add typing, that could also be a good time to drop in a couple docstrings.

JPrevost committed:
Improves logging and docstrings

Why are these changes being introduced:

* Code Review feedback

Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/GDT-122

How does this address that need:

* Adds docstrings to complex methods
* Adds useful logging, including some errors
* Fixes formating of README
* Moves list of `VALID_DOMAINS` to an ENV overridable config option

Document any side effects to this change:

* There are two known states we don't handle well still. They are noted
  by `TODO`s in the code (not a great practice) and will receive
  a Jira ticket to consider how best to handle these possible but
  unlikely states.

ghukill commented on README.md:
```diff
@@ -40,17 +40,116 @@ Not all links are publicly accessible.
+Note: See [our dev docs](https://mitlibraries.github.io/guides/authentication/touchstone_saml.html#generating-a-self-signed-certificate-for-touchstone) for information on how to generate SP key/cert. They should be unique for each deployment and backed up to our shared LastPass.
+
+- `SP_ACS_URL` = route in this app that handles the response from IDP. domain name of app + /saml/?acs
+- `SP_CERT` = obtained from self signed cert generated for this app.
+  - Note: remove all spaces/linebreaks as well as the "BEGIN" and "END" lines from file for ENV setting.
+- `SP_ENTITY_ID` = domain name of app + /saml
+- `SP_KEY` = obtained from self signed key generated for this app
+  - Note: remove all spaces/linebreaks as well as the "BEGIN" and "END" lines from file for ENV setting.
```
Thanks for adding these notes.

ghukill commented on cdnauth/auth.py:
```diff
@@ -0,0 +1,142 @@
+
+
+def prepare_flask_request(request):
+    """Return a dictionary in a format that OneLogin_Saml2_Auth uses during init"""
```
👍 to docstring

ghukill commented on cdnauth/auth.py:
```diff
@@ -72,11 +80,23 @@ def prepare_flask_request(request):
 
 @bp.route("/", methods=("GET", "POST"))
 def saml():
+    """This route handles two important situations.
+
+    The first, "sso" path, redirects a user to the IdP and sets a value of where to redirect
+    the user to after they come back to this application. "next_page" is the resource in our
+    CDN they originally requested but were redirected to this auth app because they were not
+    authenticated. Keeping track of what they wanted is essential to the functionality of
+    this solution.
+
+     The second, "acs" path is handles the response back from the IDP. We should never encounter
+     the path with "acs" in which a user is not authenticated, as they would not be redirected
+     back from the IDP to us and not be authenticated.
+    """
```
Super helpful, thanks for adding.

ghukill commented on cdnauth/auth.py:
```diff
@@ -87,8 +107,9 @@ def saml():
+            # TODO: return a useful error to the user
+            logging.error("User returned from IdP with no authentication")
+            logging.error(f"Errors: {errors}")
             pass
```
Missed this on the first review (and same for IdP response parsing error below): are we okay with a `pass` for these conditionals, where this route will then effectively return `None` by default?

I'm assuming a human wouldn't interact with this route, just the authentication handshaking... I think if a Flask route returns `None`, the actual HTTP response is a `204 - no content` response.  Assuming that's okay for whatever is interacting with this route?

If not, we could put a fall through `return` at the end of the route function with a more explicit response; though unsure if that's helpful or if we know what that response should be (e.g. `500` error + explanation for why).

JPrevost replied:
I (sort of) noted that concern myself in the latest commit message. I'll open a ticket before merging this work and link it back to this discussion thread to address this. Your suggestion to maybe fallback to some sort of error page if the more specific expected responses aren't found makes a ton of sense. I stared at it for a bit and wasn't sure how I wanted to handle it other than wanting to handle it better, so I appreciate that option being shared.

JPrevost replied:
https://mitlibraries.atlassian.net/browse/GDT-272

ghukill commented on cdnauth/config.py:
```diff
@@ -20,6 +20,10 @@ class Config:
     SP_KEY = os.getenv("SP_KEY")
     TESTING = False
     URN_UID = os.getenv("URN_UID", default="urn:oid:1.3.6.1.4.1.5923.1.1.1.6")
+    VALID_DOMAINS = os.getenv(
+        "VALID_DOMAINS",
+        default="cdn.libraries.mit.edu,cdn.dev1.mitlibrary.net,cdn.stage.mitlibrary.net",
+    )
```
👍 - seeing the associated `README` docs for this env var as well.

ghukill reviewed (approved):
Approved!  Thanks for the comments and docstrings, I think helpful for a quick footing jumping in.  

My only question -- and non blocking -- was around the `/saml` route and whether or not `pass` + default return of `None` are sufficient.  But given testing, most likely is!
